HW3/P5/label_regions.cl

```diff
@@ -80,20 +80,57 @@ propagate_labels(__global __read_write int *labels,
     old_label = buffer[buf_y * buf_w + buf_x];
 
     // CODE FOR PARTS 2 and 4 HERE (part 4 will replace part 2)
-    
+
+    /* part 2
+
+    if (old_label < w*h) {
+        buffer[buf_y * buf_w + buf_x] = labels[old_label];
+    } 
+
+    */
+
+    //part 4
+    if (lx+ly==0) { // pick (0,0) in local work group to run part 4
+        int last_ix, last_lab;
+        for (int i = 0; i < buf_w*buf_h; i++) { // loop through all elements in buffer
+            if (buffer[i] < w*h) { // if buffer element is foreground pixel:
+                if (last_ix != buffer[i]) { // is this buffer element value the same as the last one?
+                    last_ix  = buffer[i]; // save buffer element value
+                    buffer[i] = labels[buffer[i]]; // read global labels, update buffer[i]
+                    last_lab = buffer[i]; // save label
+                } else {
+                    buffer[i] = last_lab; // if buffer[i] same as previous, don't read from global to assign label
+                }
+            }
+        }
+    }
+    barrier(CLK_LOCAL_MEM_FENCE);
+
     // stay in bounds
-    if ((x < w) && (y < h)) {
+    if (((x < w) && (y < h)) && (old_label < w*h)) {
         // CODE FOR PART 1 HERE
         // We set new_label to the value of old_label, but you will need
         // to adjust this for correctness.
-        new_label = old_label;
-
+        new_label = min(
+                        old_label, //compare min of neighbors against center pixel
```

Comment:
After parts 2 and 4, you should use buffer[buf_w * buf_y + buf_x] instead of old_label.